modal/_container_entrypoint.py

```diff
@@ -114,28 +128,64 @@ async def initialize_app(self) -> _ContainerApp:
         await _container_app.init(self._client, self.app_id, self._stub_name, self._environment_name)
         return _container_app
 
-    async def _heartbeat(self):
+    async def _run_heartbeat_loop(self):
+        while 1:
+            t0 = time.monotonic()
+            if await self._heartbeat():
+                # got a cancellation event, fine to start another heartbeat soon since we know
+                # the worker supports long-polling
+                time_until_next_hearbeat = 1.0
+            else:
+                heartbeat_duration = time.monotonic() - t0
```

Comment:
If there is no cancellation *and* heartbeat returns immediately (old worker), this will ensure we wait ~15s before the next heartbeat.
If we are on a new worker however, the call will be longpolling so it wouldn't return until 15s has passed (unless there is a cancellation), so it wouldn't sleep before the next heartbeat